pkg/common/moerr/error.go

```diff
@@ -108,109 +87,112 @@ type moErrorMsgItem struct {
-	ErrMissingTxn:         {30002, 0, "missing txn"},
-	ErrUnresolvedConflict: {30003, 0, "unresolved conflict"},
-	ErrTxnError:           {30004, 0, "%s"},
-	ErrDNShardNotFound:    {30005, 0, "%s"},
+	ErrTxnClosed:          {600, 0, "the transaction has been committed or aborted"},
+	ErrTxnWriteConflict:   {601, 0, "write conflict"},
+	ErrMissingTxn:         {602, 0, "missing txn"},
+	ErrUnresolvedConflict: {603, 0, "unresolved conflict"},
+	ErrTxnError:           {604, 0, "%s"},
+	ErrDNShardNotFound:    {605, 0, "%s"},
 
 	// Group End: max value of MOErrorCode
 	ErrEnd: {65535, 0, "%s"},
 }
 
-func New(code int32, args ...any) *Error {
+func New(code uint16, args ...any) *Error {
 	return newWithDepth(Context(), code, args...)
 }
 
-func NewWithContext(ctx context.Context, code int32, args ...any) *Error {
+func NewWithContext(ctx context.Context, code uint16, args ...any) *Error {
 	return newWithDepth(ctx, code, args...)
 }
 
-func newWithDepth(ctx context.Context, code int32, args ...any) *Error {
+func newWithDepth(ctx context.Context, code uint16, args ...any) *Error {
 	var err *Error
-	item, has := errorMsgRefer[code]
-	if !has {
-		panic(fmt.Errorf("not exist MOErrorCode: %d", code))
-	}
-	if len(args) == 0 {
-		err = &Error{
-			MysqlErrCode: item.mysqlErrorCode,
-			Code:         code,
-			ErrorCode:    item.errorCode,
-			Message:      item.errorMsgOrFormat,
+	// We should try to find the corresponding error information and error code in mysql_error_define, in order to be more compatible with MySQL.
+	// you can customize moerr if you can't find it.
+	if t, ok := MysqlErrorMsgRefer[uint16(code)]; ok {
+		if len(args) == 0 {
+			err = &Error{
+				MysqlErrCode: code,
+				Code:         code,
+				ErrorCode:    0,
+				Message:      t.ErrorMsgOrFormat,
+				SqlState:     MysqlErrorMsgRefer[code].SqlStates[0],
+			}
+		} else {
+			err = &Error{
+				MysqlErrCode: code,
+				Code:         code,
+				ErrorCode:    0,
+				Message:      fmt.Sprintf(t.ErrorMsgOrFormat, args...),
+				SqlState:     MysqlErrorMsgRefer[code].SqlStates[0],
+			}
 		}
 	} else {
-		err = &Error{
-			MysqlErrCode: item.mysqlErrorCode,
-			Code:         code,
-			ErrorCode:    item.errorCode,
-			Message:      fmt.Sprintf(item.errorMsgOrFormat, args...),
+		item, has := errorMsgRefer[code]
+		if !has {
+			panic(fmt.Errorf("not exist MOErrorCode: %d", code))
+		}
+		if len(args) == 0 {
+			err = &Error{
+				MysqlErrCode: item.mysqlErrorCode,
+				Code:         code,
+				ErrorCode:    item.errorCode,
+				Message:      item.errorMsgOrFormat,
+			}
+		} else {
+			err = &Error{
+				MysqlErrCode: item.mysqlErrorCode,
+				Code:         code,
+				ErrorCode:    item.errorCode,
+				Message:      fmt.Sprintf(item.errorMsgOrFormat, args...),
+			}
 		}
 	}
+
 	_ = errors.WithContextWithDepth(ctx, err, 2)
 	return err
 }
 
 type Error struct {
 	MysqlErrCode uint16
 	ErrorCode    MOErrorCode
-	Code         int32
+	Code         uint16
```

Comment:
What is code?   Is it ErrorCode?   

Please add comments to these fields.

Reply:
Someone defines. I think it can keep only one Code, and done.